<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Why You Should Never Rescue Exception in Ruby - Daniel Fone</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link href="/atom.xml" rel="alternate" title="Daniel Fone" type="application/atom+xml">
    <link href="/favicon.png" rel="icon">
    <!-- Bootstrap -->
    <link href="/stylesheets/application.css" rel="stylesheet" type="text/css" />

    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
      <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->

    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-41009873-1', 'fone.net.nz');
  ga('send', 'pageview');

</script>

  </head>
  <body>
    <div class="blog-masthead">
      <div class="container">
        <nav class="blog-nav">
          <a href="/" class=" blog-nav-item">Home</a>
          <a href="/blog/archives/" class=" blog-nav-item">Articles</a>
          <a href="/hire/" class=" blog-nav-item">Hire</a>
          <a href="https://github.com/danielfone" class="external blog-nav-item">Code</a>
        </nav>
      </div>
    </div>

    <div class="container">
        <header>
    <img style="float: left; margin-right: 1em; width: 5em" src="/images/blue-dot.png" alt="Blue dot" />
    <h1><a href="/">Daniel Fone</a></h1>
    <p class="lead">Ruby/Rails Engineer</p>
  </header>

  <div class="row">
    <div class="col-md-9">
        <article>
    <h2>
      Why You Should Never Rescue Exception in Ruby
    </h2>
    <p>
      <small class="text-muted">
          Updated November 2014
          —
          Originally written
        May 2013
      </small>
    </p>

      <div class="well">
        <strong>tl;dr</strong> Don't write <code>rescue Exception => e</code>. Write <code>rescue => e</code> or better still, figure out exactly what you're trying to rescue and use <code>rescue OneError, AnotherError => e</code>.
      </div>

    <h3 id="whats-the-deal">What’s the deal?</h3>

<p>A common pattern for rescuing exceptions in Ruby is:</p>

<div class="highlight"><pre class="highlight ruby"><code><span class="k">def</span> <span class="nf">do_some_job!</span>
  <span class="c1"># ... do something ...</span>
  <span class="n">job_succeeded</span>
<span class="k">rescue</span>
  <span class="n">job_failed</span>
<span class="k">end</span>
</code></pre></div>
<p>This is fine, but when developers need to capture the exception details, a terrible, <em>terrible</em> thing happens:</p>

<div class="highlight"><pre class="highlight ruby"><code><span class="k">def</span> <span class="nf">do_some_job!</span>
  <span class="c1"># ... do something ...</span>
  <span class="n">job_succeeded</span>
<span class="k">rescue</span> <span class="no">Exception</span> <span class="o">=&gt;</span> <span class="n">e</span>
  <span class="n">job_failed</span> <span class="n">e</span>
<span class="k">end</span>
</code></pre></div>
<p>I have been caught out by that code on at least three separate occasions. Twice when I wrote it. I write this post in the hope that I (and perhaps others) will finally wise up about exception handling and that my fingers will never, ever type that code again.</p>

<p><em>Just to confirm this is a actually bad practice, here’s <a href="https://github.com/search?l=ruby&amp;o=asc&amp;p=1&amp;q=%22rescue+Exception+=%3E+%22&amp;ref=searchresults&amp;type=Code">~200k results</a> for <code>rescue Exception =&gt;</code> on Github</em></p>

<h3 id="what-is-this-i-dont-even">What is this I don’t even…</h3>

<p><code>Exception</code> is the root of the exception class hierarchy in Ruby. Everything from signal handling to memory errors will raise a subclass of Exception. Here’s the full list of exceptions from ruby-core that we’ll inadvertently rescue when rescuing Exception.</p>

<div class="highlight"><pre class="highlight ruby"><code><span class="no">SystemStackError</span>
<span class="no">NoMemoryError</span>
<span class="no">SecurityError</span>
<span class="no">ScriptError</span>
  <span class="no">NotImplementedError</span>
  <span class="no">LoadError</span>
    <span class="no">Gem</span><span class="o">::</span><span class="no">LoadError</span>
  <span class="no">SyntaxError</span>
<span class="no">SignalException</span>
  <span class="no">Interrupt</span>
<span class="no">SystemExit</span>
  <span class="no">Gem</span><span class="o">::</span><span class="no">SystemExitException</span>
</code></pre></div>
<p>Do you really want to rescue a <code>NoMemoryError</code> and send an email saying the job failed?!? Good luck with that.</p>

<h3 id="better-rescue-standarderror">Better: Rescue StandardError</h3>

<p><code>rescue =&gt; e</code> is shorthand for <code>rescue StandardError =&gt; e</code> and is almost certainly the broadest type of Exception that we want to rescue. In almost every circumstance, we can replace <code>rescue Exception =&gt; e</code> with <code>rescue =&gt; e</code> and be better off for it. The only time when that’s <em>not</em> a good idea is for code that’s doing some kind of exception logging/reporting/management. In those rare cases, it’s possible we’ll want to rescue non-StandardErrors — but we still need to think pretty hard about what happens after we’ve rescued them.</p>

<p>Most of the time though, we don’t even want to rescue StandardError!</p>

<h3 id="more-self-inflicted-fail">More Self-Inflicted Fail</h3>

<p>Imagine a scenario where we’re connecting to a 3rd-party API in our application. For example, we want our users to upload their cat photos to twitfaceagram. We definitely want to handle the scenarios where the connection times out, or the DNS fails to resolve, or the API returns bogus data. In these circumstances, we want to present a friendly message to the user that the application couldn’t connect to the remote server.</p>

<div class="highlight"><pre class="highlight ruby"><code><span class="k">def</span> <span class="nf">upload_to_twitfaceagram</span>
  <span class="c1"># ... do something ...</span>
<span class="k">rescue</span> <span class="o">=&gt;</span> <span class="n">e</span>
  <span class="n">flash</span><span class="p">[</span><span class="ss">:error</span><span class="p">]</span> <span class="o">=</span> <span class="s2">"The internet broke"</span>
<span class="k">end</span>
</code></pre></div>
<p>Most of the time, this code will do what we expect. Something out of our control will go wrong, and it’s appropriate to present the user with a friendly message. However, there’s a major gotcha with this code: we’re still rescuing many exceptions we’re not aware of.</p>

<p>Here’s an abridged list of StandardErrors defined in ruby-core 2.0.0 (1.9 is not materially different):</p>

<div class="highlight"><pre class="highlight ruby"><code><span class="no">StandardError</span>
  <span class="no">FiberError</span>
  <span class="no">ThreadError</span>
  <span class="no">IndexError</span>
    <span class="no">StopIteration</span>
    <span class="no">KeyError</span>
  <span class="no">Math</span><span class="o">::</span><span class="no">DomainError</span>
  <span class="no">LocalJumpError</span>
  <span class="no">IOError</span>
    <span class="no">EOFError</span>
  <span class="no">EncodingError</span>
    <span class="no">Encoding</span><span class="o">::</span><span class="no">ConverterNotFoundError</span>
    <span class="no">Encoding</span><span class="o">::</span><span class="no">InvalidByteSequenceError</span>
    <span class="no">Encoding</span><span class="o">::</span><span class="no">UndefinedConversionError</span>
    <span class="no">Encoding</span><span class="o">::</span><span class="no">CompatibilityError</span>
  <span class="no">RegexpError</span>
  <span class="no">SystemCallError</span>
    <span class="no">Errno</span><span class="o">::</span><span class="no">ERPCMISMATCH</span>
    <span class="c1"># ... lots of system call errors ...</span>
    <span class="no">Errno</span><span class="o">::</span><span class="no">NOERROR</span> <span class="c1"># errrr.... what?</span>
  <span class="no">RangeError</span>
    <span class="no">FloatDomainError</span>
  <span class="no">ZeroDivisionError</span>
  <span class="no">RuntimeError</span>
    <span class="no">Gem</span><span class="o">::</span><span class="no">Exception</span>
      <span class="c1"># ... lots of gem errors ...</span>
  <span class="no">NameError</span>
    <span class="no">NoMethodError</span>
  <span class="no">ArgumentError</span>
    <span class="no">Gem</span><span class="o">::</span><span class="no">Requirement</span><span class="o">::</span><span class="no">BadRequirementError</span>
  <span class="no">TypeError</span>
</code></pre></div>
<p>In a fresh Rails 3.2.13 application, there are <strong><a href="https://gist.github.com/danielfone/5654600">375 StandardErrors defined</a></strong>.</p>

<p>Now let’s say we’re refactoring the API integration and we make a typo with a method name. What’s going to happen?</p>

<p>If we’ve wrapped the entire process in a <code>rescue =&gt; e</code> (which is rescuing StandardError) the NoMethodError is going to be swallowed and our graceful error handling code is going to be run instead. When we run our well written tests, they’ll fail. But rather than raising a straight-forward NoMethodError, it’ll look like there was an gracefully handled connectivity problem.</p>

<p>Now <em>that</em> is going to take some debugging.</p>

<p>If our tests are poorly written there’ll be no exception and perhaps the tests will just pass. Granted, in production our users won’t be seeing ugly 500 errors, but they sure won’t be uploading their cat photos either.</p>

<h3 id="best-rescue-specific-exceptions">Best: Rescue Specific Exceptions</h3>

<p>Every part of our code is <em>qualified</em> to rescue from certain exceptional circumstances. If we want to catch connectivity problems in an API integration, our code will be qualified to rescue from a <a href="http://tammersaleh.com/posts/rescuing-net-http-exceptions">long list</a> of Net related exceptions. It is <em>not</em> qualified to rescue from an ArgumentError, which is a code-time problem and not a run-time problem!</p>

<p>Every time we write a rescue, we need to think hard about what exceptions this code is actually qualified to handle.</p>

<p>In the case of HTTP, we can make it easier on ourselves and use a wrapper like <a href="https://github.com/lostisland/faraday">faraday</a>. In this case we’ll have a <a href="https://github.com/lostisland/faraday/blob/master/lib/faraday/error.rb">much shorter list</a> of possible exceptions to rescue.</p>

<h3 id="so">So…</h3>

<p>… if you encounter <code>rescue Exception =&gt; e</code> in an existing codebase, you can almost certainly replace it with <code>rescue =&gt; e</code>.</p>

<p>… if you find yourself about to type <code>rescue Exception =&gt; e</code>, slap yourself in the face, figure out exactly what exceptions you’re dealing with and rescue those instead.</p>

  </article>

  <div id="disqus_thread"></div>
<script type="text/javascript">
    var disqus_shortname = 'danielfone';
    var disqus_identifier = 'http://daniel.fone.net.nz/blog/2013/05/28/why-you-should-never-rescue-exception-in-ruby/';

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>


    </div>
    <div class="col-md-3">
      <aside>
        <img width="70" class="img-thumbnail" style="float: right" src="//www.gravatar.com/avatar/5efc2a040a61cb22107bcfcecd58454c.jpg?s=140" alt="5efc2a040a61cb22107bcfcecd58454c" />
        <h4>About Me</h4>
        <p>
          I'm a Ruby/Rails engineer, based in Christchurch, New&nbsp;Zealand.
          I'm available for freelancing, consulting and remote contracting.
        </p>
        <p>
          Follow me on <a href="https://twitter.com/danielfone">Twitter</a>.
        </p>
        <p>
          <a href="/hire" class="button primary">Hire me</a>
        </p>
      </aside>

      <aside>
        <h4>Featured Posts</h4>
          <a href="/blog/2014/12/07/generating-unique-random-tokens/">
            <h5>Generating Unique, Random Tokens</h5>
          </a>
          <blockquote>
            If you need a random, unique token, use <code>SecureRandom.uuid</code> (or <code>SecureRandom.urlsafe_base64</code> for something shorter).
          </blockquote>
          <a href="/blog/2014/12/10/handling-token-generation-collisions-in-activerecord/">
            <h5>Handling Token Generation Collisions In ActiveRecord</h5>
          </a>
          <blockquote>
            Use <code>rescue ActiveRecord::RecordNotUnique</code> with <code>retry</code> to handle collisions when applicable.
          </blockquote>
          <a href="/blog/2013/05/24/persist-invalid-records-with-activerecord/">
            <h5>Persist Invalid Records with ActiveRecord</h5>
          </a>
          <blockquote>
            We can override ActiveRecord#save to keep invalid records in the database along with their validation errors.
          </blockquote>

        <small>
          <p><a href="/blog/archives/">More posts</a></p>
        </small>
      </aside>
    </div>
  </div>


      <div class="blog-footer">
        <p>
          &copy; 2013 &ndash;
          <script type="text/javascript">document.write(new Date().getFullYear());</script>
          Daniel Fone
          -
          Powered by
          <a href="http://middlemanapp.com/">Middleman</a>
          and
          <a href="http://pages.github.com/">GitHub Pages</a>
        </p>
        <p>
          <a href="#">Back to top</a>
        </p>
      </div>

    </div>

  </body>
</html>
